patchwork-events-world/src/main/java/net/patchworkmc/mixin/event/world/MixinServerPlayerInteractionManager.java

```diff
@@ -50,52 +45,18 @@
 
 	@Inject(method = "tryBreakBlock", at = @At("HEAD"), cancellable = true)
 	private void hookBreakBlock(BlockPos pos, CallbackInfoReturnable<Boolean> callback) {
-		boolean preCancelEvent = false;
-
-		ItemStack itemstack = player.getMainHandStack();
-
-		if (!itemstack.isEmpty() && !itemstack.getItem().canMine(world.getBlockState(pos), world, pos, player)) {
-			preCancelEvent = true;
-		}
-
-		// method_21701 => canMine
-		// Isn't the function really canNotMine?
-
-		if (player.method_21701(world, pos, gameMode)) {
-			preCancelEvent = true;
-		}
-
-		// Tell client the block is gone immediately then process events
-		if (world.getBlockEntity(pos) == null) {
-			player.networkHandler.sendPacket(new BlockUpdateS2CPacket(EmptyBlockView.INSTANCE, pos));
-		}
-
-		// Post the block break event
-		BlockState state = world.getBlockState(pos);
-		BlockEvent.BreakEvent event = new BlockEvent.BreakEvent(world, pos, state, player);
-		event.setCanceled(preCancelEvent);
-		MinecraftForge.EVENT_BUS.post(event);
-
-		// Handle if the event is canceled
-		if (event.isCanceled()) {
-			// Let the client know the block still exists
-			player.networkHandler.sendPacket(new BlockUpdateS2CPacket(world, pos));
-
-			// Update any block entity data for this block
-			BlockEntity entity = world.getBlockEntity(pos);
-
-			if (entity != null) {
-				BlockEntityUpdateS2CPacket packet = entity.toUpdatePacket();
-
-				if (packet != null) {
-					player.networkHandler.sendPacket(packet);
-				}
-			}
+		int exp = WorldEvents.onBlockBreakEvent(world, gameMode, player, pos);
 
+		if (exp < 0) {
 			callback.setReturnValue(false);
-		} else if (event.getExpToDrop() != 0) {
-			// TODO: Drop experience
-			throw new UnsupportedOperationException("Cannot drop exp from a BreakEvent yet");
+			callback.cancel();
+		} else {
+			BlockHarvestManager.pushExpDropStack(exp);
```

Comment:
doing exp stuff in this mixin is a bit confusing when the rest of it is in another mixin in a seperate module.

I understand *why* it's like this, so you don't necessarily need to change it, but I will still say it's a bit confusing.

Reply:
Exp stuff needs the world, block extension and entity package. I cannot think of a better way of doing this.
![-1b0d6ec30582ef91](https://user-images.githubusercontent.com/4704366/87462455-4f1b5d00-c653-11ea-96d8-06e88c8b6ba6.jpg)